### ISSUE TITLE
Cronet JNI crash in release and profile builds

### ISSUE THREAD

**Issue by lukehutch** [dart-lang/http#1241]
I am getting an NDK crash in `libdartjni.so (GetApplicationContext+60)` when I upload a release build of my APK to the Play Store. (The debug build runs fine locally.)

![image](https://github.com/dart-lang/native/assets/811305/e5556d34-0e5b-4693-9747-3fbe60bc5789)

I eventually ran into this additional issue of profile builds crashing locally, which I then figured out had the same cause:

https://github.com/flutter/flutter/issues/149229

This is a JNI crash, so I filed this JNI bug: https://github.com/dart-lang/native/issues/1150  However, @HosseinYousefi asked me to also file a Cronet bug, because that is the only project I am using that makes JNI calls.

Here are the Cronet setup steps I am using:

**`pubscec.yaml`**:
```yaml
  http: ^1.0.0
  cronet_http: ^1.2.1
  cupertino_http: ^1.4.0
```

**`main.dart`**:
```dart
void main() {
  // Set up Cronet http client on Android or Cupertino http client on iOS
  if (Platform.isAndroid) {
    final engine =
        CronetEngine.build(cacheMode: CacheMode.memory, cacheMaxSize: 1000000);
    runWithClient(run, () => CronetClient.fromCronetEngine(engine));
  } else if (Platform.isIOS || Platform.isMacOS) {
    final config = URLSessionConfiguration.ephemeralSessionConfiguration()
      ..cache = URLCache.withCapacity(memoryCapacity: 1000000);
    runWithClient(run, () => CupertinoClient.fromSessionConfiguration(config));
  } else {
    // Run with the default IOClient
    run();
  }
}

void run() {
  runApp(const MyApp());
}
```

I created this project as an attempt at a minimal reproducer, which adds only the above to the default Flutter project:

https://github.com/lukehutch/cronetcrash

I cannot reproduce the crash using this project, but it causes a different issue: release builds freeze on startup, before the UI is fully drawn (debug and profile builds start up fully):

<img src="https://github.com/dart-lang/http/assets/811305/a51e43c5-3322-4b93-b6b1-27d0fffe93c1" width="100"/>

**Comment by lukehutch** [comment]
Just a thought -- could the race condition be caused by the fact that I am not `await`-ing the result of `run` or `runWithClient`? Perhaps something is immediately exiting and tearing down JNI rather than waiting for its result?

```dart
Future<void> main() async {
  // Set up Cronet http client on Android or Cupertino http client on iOS
  if (Platform.isAndroid) {
    final engine =
        CronetEngine.build(cacheMode: CacheMode.memory, cacheMaxSize: 1000000);
    await runWithClient(run, () => CronetClient.fromCronetEngine(engine));
  } else if (Platform.isIOS || Platform.isMacOS) {
    final config = URLSessionConfiguration.ephemeralSessionConfiguration()
      ..cache = URLCache.withCapacity(memoryCapacity: 1000000);
    await runWithClient(run, () => CupertinoClient.fromSessionConfiguration(config));
  } else {
    // Run with the default IOClient
    await run();
  }
}

Future<void> run() async {
  await runApp(const MyApp());
}
```

**Comment by lukehutch** [comment]
I can confirm that adding `async` and `await` as shown above does prevent the lock-up on start -- as does adding `WidgetsFlutterBinding.ensureInitialized();` at the top of `main()`. So there is some race condition thing happening here.

Also, the crash happens only on the emulator (API 34, Pixel 8 Pro), not on my physical Pixel 7 Pro.

I don't know whether fixing this lockup will also fix the crash that I was getting with my full app when uploading release builds to Google Play, or when running the app in profile mode locally, but I'll keep trying to reproduce that.

**Comment by HosseinYousefi** [comment]
> as does adding `WidgetsFlutterBinding.ensureInitialized();`

I'm not an http expert but based on the docs I don't think that you can call this before `runWithClient`:

```dart
/// When used in the context of Flutter, [runWithClient] should be called before
/// [`WidgetsFlutterBinding.ensureInitialized`](https://api.flutter.dev/flutter/widgets/WidgetsFlutterBinding/ensureInitialized.html)
/// because Flutter runs in whatever [Zone] was current at the time that the
/// bindings were initialized.
```

**Comment by brianquinlan** [comment]
I don't know what the root cause is but getting Flutter applications to work with `runWithZone` is difficult. We have [updated documentation](https://github.com/dart-lang/http/blob/321362a99eeffbf83e22e5cbaa5d8356c50f73ee/pkgs/http/lib/src/client.dart#L215) that will go out with the next `package:http` release.

**Comment by HosseinYousefi** [comment]
Eureka! Maybe...

I think I have found the problem with this whole thing: `WidgetsFlutterBinding.ensureInitialized()` should not be called as stated in the docs. But if this is not initialized, then the context needed by JNI (that is filled by the plugin) is null.

I suggest not using `runWithZone`, instead inject the http dependency in another way.

fyi @lukehutch, @brianquinlan, @natebosch

**Comment by lukehutch** [comment]
> I'm not an http expert but based on the docs I don't think that you can call this before `runWithClient`:

Correct, I forgot I had seen this in the docs. In my full app, I run `ensureInitialized` within the code called by `runWithZone`. (I just tried adding it at the top of `main` in the minimal repro case to see what would happen.)

So that is probably the source of the segfault I was seeing on Google Play, if it means that the JNI context would not be initialized before cronet calls are made.

I will try the example from the docs pointed to by @brianquinlan . However if I have learned anything from tinkering with this, almost any change can resolve the crash -- there is some raciness here.

**Comment by HosseinYousefi** [comment]
> there is some raciness here.

This is due to the use of zones. I had not go into the source of `package:http` itself, so it turned out to be beneficial to open this issue in this repo.

One way to fix this problem while still using `runWithClient` is by moving ALL the client creation logic (including creation of the `CronetEngine` which required `context`) to a function to make sure it is run AFTER the initialization and not before in `main`:

```dart
if (Platform.isAndroid) {
  await runWithClient(
    run,
    () {
      final engine = 
        CronetEngine.build(cacheMode: CacheMode.memory, cacheMaxSize: 1000000);
      return CronetClient.fromCronetEngine(engine);
    },
  );
}
```

**Comment by lukehutch** [comment]
The `cronet` and `cupertino` clients should specifically point to this example code as the correct way to replace the default `http` client. 

The `http` client has an error in its docs -- it encourages initializing `cronet` or `cupertino` before Flutter is initialized:

https://pub.dev/packages/http

I don't remember where I got the `runWithClient` pattern from, but it's in some docs somewhere, and they are wrong, since Flutter can't be used with `runWithClient`...

**Comment by lukehutch** [comment]
> One way to fix this problem while still using `runWithClient` is by moving ALL the client creation logic (including creation of the `CronetEngine` which required `context`) to a function to make sure it is run AFTER the initialization and not before in `main`:

How does this ensure initialization has already happened? Are you simply adding `await`? If so, then that correlates with my other observation, that `await`-ing everything actually fixes the problem. But how does that ensure that JNI is initialized?

**Comment by lukehutch** [comment]
I started to port my code to use `provider`, but most of the most important HTTP operations use classes that do not accept an HTTP client override at time of request, such as `CachedNetworkImageProvider`.

You can provide a `CacheManager` that [creates](https://github.com/Baseflow/flutter_cached_network_image/issues/649#issuecomment-931973427) an HTTP client, but there is no `BuildContext` available when the HTTP file service is called.

So basically I really do need to get this working properly with `runWithZone` (and I suspect that a lot of other Flutter apps will too...).

**Comment by HosseinYousefi** [comment]
> How does this ensure initialization has already happened? Are you simply adding `await`? If so, then that correlates with my other observation, that `await`-ing everything actually fixes the problem. But how does that ensure that JNI is initialized?

The method that you pass to `runWithClient` will be made available as a zone variable with key `#_clientToken`.

`Client()` returns `zoneClient` if it's not null and this is the implementation of `zoneClient`:

```dart
@internal
Client? get zoneClient {
  final client = Zone.current[#_clientToken];
  return client == null ? null : (client as Client Function())();
}
```

This means that the first time `Client` is constructed, it will call the build method you passed to `runWithClient`.

Assuming all of your http requests happen after `runApp`, the first time `Client` is constructed is after context is set by Flutter.

> Should I still be wary of doing things this way, for the reasons @brianquinlan suggested

The reason I provided you with a fix for `runWithClient` is because that's the quickest way for you to change your application and get it working. And for you to know the reason behind why it failed to begin with.

**Comment by lukehutch** [comment]
> The reason I provided you with a fix for `runWithClient` is because that's the quickest way for you to change your application and get it working. And for you to know the reason behind why it failed to begin with.

Thanks. But what is the consequence of using `runWithClient` with Flutter? 

It used to be that Flutter would arbitrarily run things in the root zone, which would use the default HTTP client, not the configured client: https://github.com/dart-lang/http/issues/828  However, it looks like that was fixed. Are there still remaining issues?

**Comment by HosseinYousefi** [comment]
> Thanks. But what is the consequence of using `runWithClient` with Flutter?

I don't know enough about how Flutter deals with zones or http to give you a definitive answer, but based on what I see I don't think there should be any problems with this approach as long as you use widgets that use http inside your widget tree.

By definition, the widget tree will be created AFTER `runApp`. Say the first time `Client` gets created is inside some `CachedImage` widget, then `zoneClient` will be called and at that point the `context` is indeed available.

---

An open question is what about any changes to this context? Should the cronet engine get rebuilt after the context changes as well?  What about background http tasks? Do we have access to the same context in the background as well?

If so, then the cronet engine has to listen to context changes instead of using `Jni.getCachedApplicationContext()` and rebuild itself accordingly. To help with this `package:jni` (or really [`package:flutter_jni`](https://github.com/dart-lang/native/issues/1023) as this is flutter specific) can export some API like `addContextListener`.

**Comment by lukehutch** [comment]
Just for the record, there is one other confirmation of the JNI crash here: https://github.com/Baseflow/flutter_cached_network_image/issues/649#issuecomment-2195782869 -- so it's not just my app running on my system. But they saw this only once in their app.

**Comment by HosseinYousefi** [comment]
> Just for the record, there is one other confirmation of the JNI crash here: [Baseflow/flutter_cached_network_image#649 (comment)](https://github.com/Baseflow/flutter_cached_network_image/issues/649#issuecomment-2195782869) -- so it's not just my app running on my system. But they saw this only once in their app.

I see, I asked to know more about the way they do it. I'm personally not happy with a blanket "oh this sometimes doesn't work, so let's not use it", I want to know why it doesn't.

It might very well be an issue with the fact that cronet does not "update" the context for example.

**Comment by lukehutch** [comment]
> I see, I asked to know more about the way they do it. I'm personally not happy with a blanket "oh this sometimes doesn't work, so let's not use it", I want to know why it doesn't.

Right, exactly -- I have updated my code to not use `runWithClient`, but I still want to know why the crash was happening in the first place. Ecosystem stability always matters.

**Comment by escamoteur** [comment]
ok, I could just reproduce this, it has nothing to do with runWithClient if you use an Android without play services cronnet seems not be available. Easiest fix to prevent this crash:

```dart
const _maxCacheSize = 2 * 1024 * 1024;

Client httpClientFactory() {
  try {
    if (Platform.isAndroid) {
      final engine = CronetEngine.build(
        cacheMode: CacheMode.memory,
        cacheMaxSize: _maxCacheSize,
        enableHttp2: true,
      );
      return CronetClient.fromCronetEngine(engine);
    }

    /// 'You must provide the Content-Length' HTTP header issue
    if (Platform.isIOS || Platform.isMacOS) {
      final config = URLSessionConfiguration.ephemeralSessionConfiguration()
        ..cache = URLCache.withCapacity(memoryCapacity: _maxCacheSize);
      return CupertinoClient.fromSessionConfiguration(config);
    }
  } catch (_) {
    return IOClient(HttpClient());
  }
  var httpClient = HttpClient();
  // To use with Fiddler uncomment the following lines and set the
  // ip address of the machine where Fiddler is running
  // httpClient.findProxy = (uri) => 'PROXY 192.168.1.61:8866';
  // httpClient.badCertificateCallback =
  //     (X509Certificate cert, String host, int port) => true;
  return IOClient(httpClient);
}
```

**Comment by escamoteur** [comment]
@brianquinlan propbably this precaution should be added to the example and the HTTP docs

**Comment by lukehutch** [comment]
@escamoteur The crash I was seeing was on Google Experience / Google Play Services devices (Pixel Pro 7 physical device, and Pixel Pro 8 emulator). But thanks for pointing out the need for `try...catch`.

**Comment by escamoteur** [comment]
@lukehutch actually your crash on top here seems to be a different one than the JNI one

**Comment by escamoteur** [comment]
I got this here in that case:

```
java.lang.RuntimeException: All available Cronet providers are disabled. A provider should be enabled before it can be used.
	at org.chromium.net.CronetEngine$Builder.getEnabledCronetProviders(CronetEngine.java:8)
	at org.chromium.net.CronetEngine$Builder.createBuilderDelegate(CronetEngine.java:3)
	at org.chromium.net.CronetEngine$Builder.<init>(CronetEngine.java:1)
  File "accessors.dart", line 119, in JniAccessorWrappers.throwException
  File "accessors.dart", line 11, in _check
  File "accessors.dart", line 16, in JniResultMethods.check
  File "accessors.dart", line 54, in JniResultMethods.objectPointer
  File "accessors.dart", line 59, in JniResultMethods.reference
  File "jni_bindings.dart", line 2317, in new CronetEngine_Builder
  File "cronet_client.dart", line 97, in CronetEngine.build
  File "http_client_factory.dart", line 12, in httpClientFactory
  File "zone.dart", line 1399, in _rootRun
  File "zone.dart", line 1656, in _RootZone.run
  File "zone.dart", line 1616, in _RootZone.bindCallback.<fn>
  File "client.dart", line 166, in zoneClient
  File "client.dart", line 42, in new Client
  File "di.dart", line 86, in _registerBaseScope
  File "di.dart", line 78, in setupDi
  File "app.dart", line 76, in _AppWidget.build.<fn>.<fn>.<fn>.<fn>
  File "interaction_manager_impl.dart", line 357, in InteractionConnectorState.initState
  File "framework.dart", line 5618, in StatefulElement._firstBuild
```

**Comment by escamoteur** [comment]
@lukehutch how do you link the NDK stack trace to the JNI call?
does it also happen when you don't use runWithClient but provide the Client via Provider or GetIt?

This is what we do to setup the cacheManager

```dart
  const cacheKey = 'watchcrunch_cache';
  di.registerSingleton(Client());
  final appState = di<AppState>();
  di.registerSingleton<CacheManager>(
    WcImageCacheManager(
      Config(
        cacheKey,
        fileService: WcHttpFileService(
          httpClient: di<Client>(),
        ),
        stalePeriod: appState.imageDiscStalePeriod,
        maxNrOfCacheObjects: appState.imageDiscCacheSize,
      ),
    ),
```

and we do the runWithClient directly in our main and we don't await it. Not problem so far with our App that we released yesterday with the new Clients. only the one without Play Services

**Comment by lukehutch** [comment]
> @lukehutch how do you link the NDK stack trace to the JNI call?

The top line of the stacktrace is in `libdartjni.so`...

> does it also happen when you don't use runWithClient but provide the Client via Provider or GetIt?

So far, no. I am not using `runWithClient` anymore though just to be safe, I just explicitly use the `Client` instance every time I need to make an HTTPS request.

It sounds like @HosseinYousefi has a good understanding of what was causing the crash. The solution might just be better documentation, but I would hope there is a way to more robustly protect against JNI being uninitialized when it's needed. The open questions in https://github.com/dart-lang/http/issues/1241#issuecomment-2195765525 will hopefully be answered before this is closed.

**Comment by escamoteur** [comment]
hmm, what I don't understand is why it works without any problems for us.
btw we use newer versions of HTTP and the clients

  cronet_http: ^1.3.1
  cupertino_http: ^1.5.0

dependency_overrides:
  http: ^1.2.0

**Comment by brianquinlan** [comment]
If Cronet is not available then you should get a clear
`java.lang.RuntimeException: All available Cronet providers are disabled. A provider should be enabled before it can be used.` exception.

The [`package:cronet_http` documentation](https://pub.dev/packages/cronet_http#use-embedded-cronet) explains how to get the package to work if the phone does not include Google Play Services.

**Comment by lukehutch** [comment]
> The [`package:cronet_http` documentation](https://pub.dev/packages/cronet_http#use-embedded-cronet) explains how to get the package to work if the phone does not include Google Play Services.

Thanks for the tip, Brian. However, is there no way to do this without using a commandline switch? Shouldn't it be possible to depend upon `org.chromium.net:cronet-embedded` directly? And/or maybe a package `cronet_http_embedded` could be created that only depends on the embedded version?

Are there any downsides to using the embedded version? If not, what benefit is obtained by requiring Play Services for the non-embedded version?

**Comment by escamoteur** [comment]
The downside is that cronnet is bundled with the apk in that case which increases your app's size. It doesn't seem possible to download it automatically when it isn't bundled or play services are not available.

@brianquinlan it probably would be helpful if the
exception would mention the dependency on Google Play Services and/or the docs on the configuration of the http package would contain a sentence mentioning it (the table about which platform supports which clients just states Android)
Also the given example code should contain a try catch to handle the situation that no cronnet is available.
Am 29. Juni 2024, 04:42 +0100 schrieb Luke Hutchison ***@***.***>:

**Comment by HosseinYousefi** [comment]
This should not be any different than using cronet on Android without Flutter. How do they solve this?

This beginner codelab seems to have the answer: https://developer.android.com/codelabs/cronet#4

They use `CronetProviderInstaller.installProvider`. What if we expose this API and add similar documentation? Then if the play services are not available, the users can install it automatically.

**Comment by escamoteur** [comment]
@HosseinYousefi if that is possible with flutter it should be the way to go. Ideally the package should take care of this on his own.

**Comment by lukehutch** [comment]
Why can't pre-built libraries be shipped with the Flutter package, like with every other NDK library? Is there a good reason the package has to be fetched from Google Play?

**Comment by HosseinYousefi** [comment]
> @HosseinYousefi if that is possible with flutter it should be the way to go. Ideally the package should take care of this on his own.

I don't see why it shouldn't be possible. Feel free to make a PR

> Why can't pre-built libraries be shipped with the Flutter package, like with every other NDK library? Is there a good reason the package has to be fetched from Google Play?

You can ship it. However there are some benefits that you gain if you use play services. From https://developer.android.com/codelabs/cronet#3:

> The Cronet team recommends using the Google Play Services provider. By using the Google Play Services provider, your application doesn't need to pay the binary size cost of carrying Cronet (about 5 megabytes) and the platform ensures that latest updates and security fixes are delivered.

Simply follow the instructions here: https://github.com/dart-lang/http/tree/master/pkgs/cronet_http#use-embedded-cronet

**Comment by escamoteur** [comment]
Yeah, 5MB is quite a lot. Not sure how much is saved if you select to exclude the default HttpClient implementation.
Downloading on demand would be the best.
Unfortunately I m no Android native dev, so someine else would have to make that PR
Am 30. Juni 2024, 11:53 +0100 schrieb Hossein Yousefi ***@***.***>:

**Comment by HosseinYousefi** [comment]
> Unfortunately I m no Android native dev, so someine else would have to make that PR

Yeah, I didn't mean you specifically –

However we're making it really easy to develop these things using JNIgen. Anyone implementing this wouldn't have to write any Java or Kotlin code. Everything can be done using the generated Dart bindings.

**Comment by lukehutch** [comment]
> You can ship it. However there are some benefits that you gain if you use play services

Could we get a cronet_http_embedded package on pub.dev though, please, for developers who just want to ship it with their app? For Play experience devices, the total download size is the same. I'm not worried about adding another 5MB when tons of apps today have download sizes over 100MB.

**Comment by brianquinlan** [comment]
> This should not be any different than using cronet on Android without Flutter. How do they solve this?
> 
> This beginner codelab seems to have the answer: https://developer.android.com/codelabs/cronet#4
> 
> They use `CronetProviderInstaller.installProvider`. What if we expose this API and add similar documentation? Then if the play services are not available, the users can install it automatically.

I think that this will only work if the device has Play Services installed. The last line says:

> This launches a Play Services [Task](https://developers.google.com/android/guides/tasks) which installs the provider asynchronously.

**Comment by lukehutch** [comment]
> Could you explain why the configuration flag doesn't work for you?

I couldn't understand how to use this. Is this a runtime flag, or a compiletime flag? The examples are for `flutter run` and `flutter test`, not `flutter build`, so it's not clear. If it's a runtime flag, how do you get a deployed app to run with the equivalent of that flag at launch? If it's a compiletime flag, I haven't been able to figure out how to add flags to my build...

It's all good though, I changed my code to just fall back to `IOClient` if initializing Cronet throws an exception... so it should work on all devices, at least.

The things I would still love to see are:
1. `try...catch` added to the examples in the documentation, with a fallback to `IOClient` on exception.
2. Some resolution to the questions in https://github.com/dart-lang/http/issues/1241#issuecomment-2195765525
3. Ideally, a check in Cronet to see if JNI has been initialized before trying to make JNI calls (so that your app doesn't inexplicably crash).

I had another weird lockup issue on startup (I'll link it below), and I think it was also due to the JNI initialization issue, because as far as I can tell (it was sporadic), it went away when I stopped using `runWithClient`. I could never figure out what was going on because no debugger would attach. I just now realized it's the same issue as the freezing issue that I ran into with my minimal reproduction testcase for this issue. It would be nice to completely avoid this kind of problem by having some more robust behavior in Cronet or JNI to prevent these sorts of issues.

**Comment by brianquinlan** [comment]
> > Could you explain why the configuration flag doesn't work for you?
> 
> I couldn't understand how to use this. Is this a runtime flag, or a compiletime flag? The examples are for `flutter run` and `flutter test`, not `flutter build`, so it's not clear. If it's a runtime flag, how do you get a deployed app to run with the equivalent of that flag at launch? If it's a compiletime flag, I haven't been able to figure out how to add flags to my build...
> 

It's a compile time flag. You can use that same flag with the `flutter build` subcommands.

**Comment by brianquinlan** [comment]
My point was that you still need Google Play Services. If you don't have Google Play Services installed (e.g. you need to support a phone distributed in China) then this approach won't work.

**Comment by lukehutch** [comment]
OK, thanks Brian and Houssein for explaining the Google Play situation. I'll leave this issue open for the other JNI-related stuff (I believe it is still important to avoid crashes and lockups even if developers try to initialize Cronet before Flutter -- and the docs need improving).

**Comment by mkustermann** [comment]
> I don't know enough about how Flutter deals with zones or http to give you a definitive answer, but based on what I see I don't think there should be any problems with this approach as long as you use widgets that use http inside your widget tree.

The zone machinery works as long as one uses it as it was designed: If a library accepts callback functions, the library should bind those callback functions to the current zone, ensuring that once the callbacks are invoked (caller can be from another zone) it switches to the target zone (where the callback was bound to) before executing the callback.

In pure Dart this works, because Dart's core libraries ensure callbacks are bound to the current zone (e.g. `Future`, `Stream`, `ReceivePort`, ...). For example

```dart
import 'dart:async';

void main() {
  late final Zone fromZone;
  late final Zone toZone;
  late final Stream stream;

  runZoned(() async {
    fromZone = Zone.current;
    final controller = StreamController();
    stream = controller.stream;
    for (int i = 0; i < 10; ++i) {
      await Future.delayed(const Duration(milliseconds: 1));
      controller.add(i);
    }
  });
  
  runZoned(() async {
    toZone = Zone.current;
    stream.listen((data) {   // <-- this will use `Zone.current.bindUnaryCallback()` internally
      print('Adding zone is different from consuming zone: ${!identical(fromZone, toZone)}');
      print(data);
    });
  });
}
```

Here the callback given to `stream.listen((data) {})` will be bound to the current zone in the implementation of `stream.listen()`, ensuring the zone is switched to the correct zone when invoking the callback given to `listen()`.

=> This only works because `Stream`, `Future`, `dart:io` APIs, ... use `Zone.current.bind*Callback()` on functions they accept, thereby ensuring they'll be invoked in the right zone.

Many of flutter's APIs do not do that, for example taking the default `flutter create` counter app and modifying it like this:
```diff
class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

+  late Zone expectedZone;

  void _incrementCounter() {
+    print('Button pressed callback in correct zone: ${identical(expectedZone, Zone.current)}'); 
+    print('Button pressed callback in run in root zone: ${identical(Zone.root, Zone.current)}');

    setState(() { _counter++; });
  }

  @override
  Widget build(BuildContext context) {
+    return runZoned(() {
+      expectedZone = Zone.current;
        return ... (
          floatingActionButton: FloatingActionButton(
            onPressed: _incrementCounter,
            tooltip: 'Increment',
            child: const Icon(Icons.add),
          ...);
+    });
  }
}
```
This will print
```
flutter: Button pressed callback in correct zone: false
flutter: Button pressed callback in run in root zone: true
```

=> The callback given to `FloatingActionButton` is not invoked in the correct zone.
=> This indicates `FloatingActionButton` doesn't use `Zone.current.bindCallback(<>)`

We can "fix" this manually via
```diff
        floatingActionButton: FloatingActionButton(
-          _incrementCounter,
+          onPressed: Zone.current.bindCallback(_incrementCounter),
          ...);
```

=> So the root cause is flutter's choice of not binding callbacks obtained from APIs to the current zone.


Flutter under the hood has a C++ engine, that will decide when frames are to be drawn and explicitly call into Dart to render frames. It has to choose what zone those calls should run inside of.

=> I believe it will use the zone that was active when the bindings were initialized.

For every frame to be drawn flutter C++ engine will call [_drawFrame](https://github.com/flutter/engine/blob/cc24bbb859c9ef200d537577a608707e4d84e772/lib/ui/hooks.dart#L282)
```dart
@pragma('vm:entry-point')
void _drawFrame() {
  PlatformDispatcher.instance._drawFrame();
}
```
which calls [PlatformDispatcher.instance._drawFrame](https://github.com/flutter/engine/blob/cc24bbb859c9ef200d537577a608707e4d84e772/lib/ui/platform_dispatcher.dart#L418)
```dart
  // Called from the engine, via hooks.dart
  void _drawFrame() {
    _invoke(onDrawFrame, _onDrawFrameZone);
  }
```
which will call `onDrawFrame` in the `_onDrawFrameZone` zone.

=> That zone is the one that was used when initializing the bindings in flutter, which either happens implicitly in `runApp` somewhere or explicitly via e.g. calls to `WidgetsFlutterBinding.ensureInitialized()` and similar.
=> So anything it calls from there, widget building etc will use that zone
=> For pointer events / clicks / ... it probably follows a very similar flow, and eventually invokes callbacks given to widget constructors, which - as they were not bound to the zone at widget constructor time - will just run in the zone that the bindings were setup

**Comment by mkustermann** [comment]
> I think I have found the problem with this whole thing: `WidgetsFlutterBinding.ensureInitialized()` should not be called as stated in the docs. But if this is not initialized, then the context needed by JNI (that is filled by the plugin) is null.

@HosseinYousefi Two questions:
a) Wouldn't we throw a nice exception if jni related code requires an android context object and it wasn't initialized? (It shouldn't crash, hang or otherwise - it should provide a nice error message explaining the problem and how to address it)
b) What would it take to make android context objects available to JNI code without having to rely on flutter bindings being initialized? (Tangential: Also thinking about `Isolate.run()` / `compute()` running things in helper isolates which may not have bindings being setup)

**Comment by HosseinYousefi** [comment]
> a) Wouldn't we throw a nice exception if jni related code requires an android context object and it wasn't initialized? 

Android context can be nullable. Also there might be a need for listening to context changes, it's good to expose some APIs that enable this.
https://github.com/dart-lang/native/issues/1254

> b) What would it take to make android context objects available to JNI code without having to rely on flutter bindings being initialized? (Tangential: Also thinking about `Isolate.run()` / `compute()` running things in helper isolates which may not have bindings being setup)

Yes, we're relying on the existing Flutter plugin system. I will explore this when working on "`package:flutter_jni`".
https://github.com/dart-lang/native/issues/1255

Maybe cronet specifically doesn't need the context from Flutter plugin system and can instead use the global context. That one is easy to get: `ActivityThread.currentActivityThread().getApplication()`

**Comment by brianquinlan** [comment]
> The zone machinery works as long as one uses it as it was designed...

I think that using `runWithClient` is too hard to get right in Flutter, which is why I'd suggest using a state-management solution like `package:providers`.

**Comment by escamoteur** [comment]
actually we use runWithClient in our app without any problems with Fluter. I'm pretty sure the reason for this sort of crash is something else

**Comment by brianquinlan** [comment]
@lukehutch are you running on 32-bit Android? If so, there was a bug fixed in `package:cronet_http` 1.3.1. In any case, could you try to reproduce in the last version?

**Comment by HosseinYousefi** [comment]
> actually we use runWithClient in our app without any problems with Fluter. I'm pretty sure the reason for this sort of crash is something else

I identified the reason. It was the fact that context was not initialized. You can find my answer above. 

I already mentioned the 32 bit bug.

**Comment by HosseinYousefi** [comment]
@brianquinlan We can safely close this issue. Here's why the crash happened: https://github.com/dart-lang/http/issues/1241#issuecomment-2195694282

**Comment by brianquinlan** [comment]
Do you think that it is worth mentioning in the docs that either `runApp` or `ensureInitialized` should be called before making any calls to `cronet_http`?

**Comment by HosseinYousefi** [comment]
> Do you think that it is worth mentioning in the docs that either `runApp` or `ensureInitialized` should be called before making any calls to `cronet_http`?

Yes, that's helpful to mention. I have opened two issues to 1. Output a descriptive error message if users do this. 2. See if I can stop depending on flutter bindings for retrieving the right context needed by cronet. I'll open PRs here if to upgrade the dependencies once that's done as usual!

**Comment by brianquinlan** [comment]
Do you think that work will be done in the short term? I don't want to add documentation if I'll end it removing it (assuming that I don't forget!) in a month.

**Comment by HosseinYousefi** [comment]
> Do you think that work will be done in the short term? I don't want to add documentation if I'll end it removing it (assuming that I don't forget!) in a month.

I'll prioritize no. 1, but that's in line with the documentation.

I don't know even if no. 2 is possible right now. I have to check.

**Comment by zigapovhe** [comment]
@HosseinYousefi Please add this to the docs! We lost few days before coming up to this issue. The worse thing is that only on a few of our devices the app kept crashing on app startup.  Moving init of cronet_http after runApp() solved the issue. (Before that it was right under WidgetsFlutterBinding.ensureInitialized(); and it was crashing for some users.

![Image](https://github.com/user-attachments/assets/f30911c1-180a-49a6-8e58-b2db4ee69dad)

**Comment by HosseinYousefi** [comment]
> [@HosseinYousefi[<img alt="" width="15" height="15" src="chrome-extension://dlebflppeeemcdpidccbiblndppbmjmh/ospo-chrome-ext-logo.png">](https://teams.googleplex.com/yousefi@google.com)](https://github.com/HosseinYousefi) Please add this to the docs! 

It is in the docs: https://github.com/dart-lang/http/blob/master/pkgs/cronet_http/lib/cronet_http.dart#L9-L13

**Comment by zigapovhe** [comment]
Oh my bad. In my case, WidgetsFlutterBinding.ensureInitialized(); was right above cronet_http initialisation and it still crashed from time to time. It would be beneficial if this would also be written in README.md so it can be seen on pub.dev as well.
